### PR TITLE
Issue 4817: Updated the engine coding guidelines

### LINKED ISSUE #4817: Update the engine programming guidelines
Update the programming guide lines to reflect the guidelines and style that we use.

### PR DESCRIPTION
**Pull request by JCash**

### CODE REVIEW

File: README_ENGINE.md [hunk 1]
```diff
@@ -37,55 +45,204 @@ The engine source code can be found in the `/engine` folder. The source code is
...
 
+</details>
+
+## Code
+
+One of our goals is to keep the engine as small and performant as possible.
+By this we mean both development time, and actual runtime for the released game:
+compiled code size, source code size, compile time, runtime.
+
+By thinking this in mind whenever we add new features, we can keep this goal with minimal effort.
```
Comment (eangelin) [inline README_ENGINE.md:56]: keeping

File: README_ENGINE.md [hunk 2]
```diff
@@ -37,55 +45,204 @@ The engine source code can be found in the `/engine` folder. The source code is
...
+The fastest code is the code that isn't run, or isn't even added!
+
+If we don't need it, don't add it.
+
+Think of _all_ the end users. Will the fix/feature have a clear benefit to most of our users?
+If not, can the feature be added as a separate native extension instead?
+
+### Design
+
+Before adding code to the engine, the code needs a design first.
```
Comment (eangelin) [inline README_ENGINE.md:69]: I suggest adding this as a "#### Features" section, to emphasize that for new features we really want to see a design (and maybe even a community discussion first). For Fixes the bar is obviously lower.

Comment (eangelin) [reply README_ENGINE.md:69]: ... then we can also say "before adding a feature to the engine, the feature...."

File: README_ENGINE.md [hunk 3]
```diff
@@ -37,55 +45,204 @@ The engine source code can be found in the `/engine` folder. The source code is
...
+
+Design problems caught in a design review is a huge time saver, and it also improves everyones understanding of the code and problem area.
```
Comment (eangelin) [inline README_ENGINE.md:71]: Suggesting "Catching problems in a design review is..."

File: README_ENGINE.md [hunk 4]
```diff
@@ -37,55 +45,204 @@ The engine source code can be found in the `/engine` folder. The source code is
...
+
+#### Fixes
+
+For the simple bug fixes, it's usually enough to add a pull request, with little to no design phase beforehand.
+
+Comment the pull request well, and the use case(s) it fixes.
```
Comment (eangelin) [inline README_ENGINE.md:77]: Maybe "issues it fixes", and ask them to link to the issue?

File: README_ENGINE.md [hunk 5]
```diff
@@ -37,55 +45,204 @@ The engine source code can be found in the `/engine` folder. The source code is
...
+Each solution should have its pros and cons listed, so that reviewers can reason about them.
+
+
+* Keep it short
+	* The document doesn't need to be fancy, or long. Aim for 1-2 pages, as long as it is easy to read and understand for the reviewers.
+
+
+* Keep it on point
+	* The design should only deal with the actual problem. It is easy to think in too generic terms.
+The design review will also help highlight such issue.
```
Comment (eangelin) [inline README_ENGINE.md:91]: issues

File: README_ENGINE.md [hunk 6]
```diff
@@ -37,55 +45,204 @@ The engine source code can be found in the `/engine` folder. The source code is
...
+
+Although we recommend starting with a design review, to get a first go-ahead, it is sometimes required to do some test
+code to see if the idea pans out or not. Be aware that the community might have already touched this idea beforehand
+and discarded it for one reason or another. So try to ask first, before spending too much time implementating the feature.
+
+*Note: we still don't have a good shared place to store the design documents, where users can comment on it.
+Our current best recommendation is to share a google drive document. If yoiu have a good alternative, let us know!*
```
Comment (eangelin) [inline README_ENGINE.md:98]: you

File: README_ENGINE.md [hunk 7]
```diff
@@ -37,55 +45,204 @@ The engine source code can be found in the `/engine` folder. The source code is
...
+
+#### Backwards compatibility
+
+Even for the least complex fixes, there might be nuances that aren't obvious at first glance.
+Be aware that there are many projects out there, and we expect them to be able to build with new versions of Defold.
```
Comment (eangelin) [inline README_ENGINE.md:103]: scratch "to be able"?

File: README_ENGINE.md [hunk 8]
```diff
@@ -37,55 +45,204 @@ The engine source code can be found in the `/engine` folder. The source code is
...
+
+Sometimes we can argue it's a bug fix, and we go ahead with the implementation.
```
Comment (eangelin) [inline README_ENGINE.md:105]: Maybe "Sometimes we can argue a breaking change is a bug fix"

File: README_ENGINE.md [hunk 9]
```diff
@@ -37,55 +45,204 @@ The engine source code can be found in the `/engine` folder. The source code is
...
 
-## No C++ Exceptions
 We don’t make use of any exceptions in the engine. It isn’t generally used in game engines, since the data is (mostly) known beforehand, during development. Removing the support for C++ exceptions decreases executable size and improves the runtime performance.
 
+#### No STL
+
+While it might be tempting to use `std` functions/types/classes, we don't do it for several reasons.
+Performance is one (code size, compilation time, debugging time), and ABI issues is another important issue.
+
+You can read abit more about the advice we give to extension developers here:
```
Comment (eangelin) [inline README_ENGINE.md:148]: a bit

File: README_ENGINE.md [hunk 10]
```diff
@@ -37,55 +45,204 @@ The engine source code can be found in the `/engine` folder. The source code is
...
+https://defold.com/manuals/extensions-best-practices/#standard-template-libraries---stl
+
+We do make a (very) few exceptions, most notable for `std::sort()` and `std::upper_bound()`.
```
Comment (eangelin) [inline README_ENGINE.md:151]: notably

File: README_ENGINE.md [hunk 11]
```diff
@@ -37,55 +45,204 @@ The engine source code can be found in the `/engine` folder. The source code is
...
+Sometimes we need to use a third party library.
+It is fine, if the cost of using it (code size, complexity, performance) is properly assessed.
+Also make sure to list alternatives (such as implementing the feature yourself) in the design phase.
+
+#### Debug vs Release
+
+We don't really distinguish between the traditional Debug or Release concept.
+We compile (and ship) all our libraries with the `-O2` flag.
+
+Instead, we compartmentalize functionality into separate libraries, and during link time, either add `featureX_impl.a` or a `featureX_null.a` to the command line.
```
Comment (eangelin) [inline README_ENGINE.md:194]: Should we say this is how we handle platform differences as well? feature_win32 feature_darwin? and #ifdefs if it's a small difference?

File: README_ENGINE.md [hunk 12]
```diff
@@ -13,9 +21,9 @@ The engine source code can be found in the `/engine` folder. The source code is
 * **facebook** - Facebook stub implementation. Actual implementation moved to [extension-facebook](https://github.com/defold/extension-facebook)
 * **gameobject** - Game object and component lifecycle
 * **gamesys** - Component and resource types are registered here
-* **glfw** - Somewhat modified version of GLWF, used to set up a window/graphics context and read from input devices. Defold uses an old GLFW 2.7 version with some features backported from GLFW 3.0.
```
Comment (rgrams) [inline README_ENGINE.md:16]: GLFW

**Review by eangelin — DISMISSED** [review]

**Pull request merged**